modules/container/container.go

```diff
@@ -0,0 +1,219 @@
+package container
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"syscall"
+	"time"
+
+	"github.com/opencontainers/runtime-spec/specs-go"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/cio"
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/oci"
+	"github.com/google/shlex"
+	"github.com/threefoldtech/zbus"
+	"github.com/threefoldtech/zosv2/modules"
+	"github.com/threefoldtech/zosv2/modules/stubs"
+)
+
+const (
+	containerdSock = "/run/containerd/containerd.sock"
+)
+
+type containerModule struct {
+	flister    *stubs.FlisterStub
+	containerd string
+}
+
+func New(zcl zbus.Client, containerd string) modules.ContainerModule {
+	if len(containerd) == 0 {
+		containerd = containerdSock
+	}
+
+	return &containerModule{
+		flister:    stubs.NewFlisterStub(zcl),
+		containerd: containerd,
+	}
+}
+
+func getNetworkSpec(network modules.NetworkInfo) oci.SpecOpts {
+	ns := network.Namespace
+	if !path.IsAbs(ns) {
+		// just name
+		ns = path.Join("/var/run/netns", ns)
+	}
+
+	return oci.WithLinuxNamespace(
+		specs.LinuxNamespace{
+			Type: specs.NetworkNamespace,
+			Path: ns,
+		},
+	)
+}
+
+func withHooks(hooks specs.Hooks) oci.SpecOpts {
+	return func(_ context.Context, _ oci.Client, _ *containers.Container, spec *oci.Spec) error {
+		spec.Hooks = &hooks
+		return nil
+	}
+}
+
+// NOTE:
+// THIS IS A WIP Create action and it's not fully implemented atm
+func (c *containerModule) Run(ns string, name string, flist string, tags, env []string, network modules.NetworkInfo,
+	mounts []modules.MountInfo, entrypoint string) (id modules.ContainerID, err error) {
+	// create a new client connected to the default socket path for containerd
+	client, err := containerd.New(c.containerd)
+	if err != nil {
+		return id, err
+	}
+	defer client.Close()
+
+	args, _ := shlex.Split(entrypoint)
+
+	root, err := c.flister.Mount(flist, "")
+	if err != nil {
+		return id, err
+	}
+
+	defer func() {
```

Comment:
You should not defer Umount here.
You only unmount when you delete the container.

Reply:
we defer here only if an error occurs (container failed to start) so there will not be a delete call afterwards

Reply:
I'm really not sure about this pattern.
I think defer does evaluate the the value of err when you call defer, not when the function deferred is called.
https://golang.org/doc/effective_go.html#defer
So this means here err will always be false at this point, thus the Umount will never bee called even in the case there is an error when you create the container.

Reply:
I don't believe so, i used this pattern before many times and it works as expected. but let me test again to be sure.

Reply:
Notice also i use a named return. so (err) in all function reference the same err object, hence once the err is actually populated to value it will be detected by error.

Reply:
https://play.golang.org/p/_sKT9UBDwdV